core/src/edu/cornell/gdiac/main/CollisionController.java

```diff
@@ -1,32 +1,132 @@
+import com.badlogic.gdx.physics.box2d.Manifold;
+import com.badlogic.gdx.utils.ObjectSet;
+import edu.cornell.gdiac.physics.obstacle.BoxObstacle;
+import edu.cornell.gdiac.physics.obstacle.Obstacle;
 
-public class CollisionController {
+public class CollisionController implements ContactListener {
 
-  /* TODO P5 add fields that you may need
-  This could include the level, avatar, bounce pad, etc..
-  If you notice that after adding these the gameplay controller doesn't need it, remove it from the
-  gameplay controller.
-
-  TODO P5 Also when you are handling collisions, make sure that you don't directly modify game controller values.
-  For example for detecting level completion, we could have an "is complete" flag in the level model,
-  updating this and then the GameController reads from that to avoid cyclic dependencies.
+  /**
+   * Mark set to handle more sophisticated collision callbacks
    */
-  private LevelModel levelModel;
+  protected ObjectSet<Fixture> sensorFixtures;
+  private LevelModel level;
+  private boolean complete;
+  private float bounceForce = 10f;
 
   public CollisionController(LevelModel levelModel) {
-    this.levelModel = levelModel;
+    this.level = levelModel;
+    this.sensorFixtures = new ObjectSet<Fixture>();
+  }
+
+  /**
+   * Return the complete flag to GameController
+   *
+   * @return
+   */
+  public boolean isComplete() {
+    return complete;
+  }
+
+  /**
+   * Change the complete flag to true if player touches the goal
+   *
+   * @param value
+   */
+  public void setComplete(boolean value) {
+    complete = value;
   }
 
-  public void handleContact(Contact contact) {
-    // TODO P5 this method should call other collision methods based on which two objects are colliding
-    // P5 should work on getting this the default collisions to work using this method.
+  public void beginContact(Contact contact) {
+    Fixture fix1 = contact.getFixtureA();
+    Fixture fix2 = contact.getFixtureB();
+
+    Body body1 = fix1.getBody();
+    Body body2 = fix2.getBody();
+
+    Object fd1 = fix1.getUserData();
+    Object fd2 = fix2.getUserData();
+
+    try {
+      Obstacle bd1 = (Obstacle) body1.getUserData();
+      Obstacle bd2 = (Obstacle) body2.getUserData();
+
+      PlayerModel avatar = level.getAvatar();
+      BoxObstacle door = level.getExit();
+
+      // Check if any fixture is a bounce platform
+      BouncePlatformModel bouncePlatform = null;
+      if (bd1 instanceof BouncePlatformModel) {
+        bouncePlatform = (BouncePlatformModel) bd1;
+      } else if (bd2 instanceof BouncePlatformModel) {
+        bouncePlatform = (BouncePlatformModel) bd2;
+      }
+
+      // See if we have landed on the ground.
+      if ((avatar.getSensorName().equals(fd2) && avatar != bd1) ||
+          (avatar.getSensorName().equals(fd1) && avatar != bd2)) {
+        avatar.setGrounded(true);
+        sensorFixtures.add(avatar == bd1 ? fix2 : fix1); // Could have more than one ground
```

Comment:
Wdym by the comment "Could have more than one ground"?

Reply:
I think that is in the original code, I am not sure about it too :( I suppose it means that we could theoretically have the player touching two ground at the same time?

Reply:
lol I guess so, if it was in the original code I guess we can just leave it for now.